### PR TITLE
Add a post: Fine-Tuning for Semantic Parsing with Levanter

### PR DESCRIPTION
**Pull request by Ivan-Zhou**
First PR of 2024! 

In this PR, I added a post on how to fine-tune Llama2-7B on GEM ViGGO using Levanter. I compared both full-weight fine-tuning and LoRA fine-tuning, with the out-of-box Llama2-7B Chat. 

Head:
image

Results:
image

### CODE REVIEW

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+gcloud compute tpus tpu-vm ssh finetune-32 --zone us-east1-d --worker=all \
+--command="WANDB_API_KEY=${YOUR WANDB TOKEN HERE} \
+HUGGING_FACE_HUB_TOKEN=${YOUR HF TOKEN HERE} \
+bash levanter/infra/run.sh python \
+levanter/examples/alpaca/alpaca.py \
+--config_path gs://<config-yaml-file> \
+--hf_save_path gs://<somewhere>"
+```
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:138]: 🐎

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+Below shows an example where we compare the output from the original chat model and the fine-tuned model. We can see that the fine-tuned model is able to generate the correct function call and precisely capture the right attributes, while the original model is not able to do so.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:140]: maybe give some numbers (baseline vs fine tuned) right before the example, just to set some expectations. Typical flow for something like this is `task overview -> dataset overview -> approach -> baseline results -> (exp -> results -> error analysis/example)*`

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:144]: can you maybe format so that the outputs start at the same column so it's easy to visually see the comparison

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+Generation from Llama2-7B Chat: \nOutput: inform(name[SpellForce 3], release_year[2017], esrb[NA], genres[fantasy,\naction, adventure], platforms[PC], available_on_steam[no], has_linux_release[no],\nhas_mac_release[no])
+Generation from fine-tuned Llama2-7B: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+```
+
+#### LoRA Fine-tuning
+
+Below is our configuration for LoRA fine-tuning. Note that it is very similar to the full-weight fine-tuning configuration, except for a few differences:
+
+- We increased the number of steps by 1 more epoch. LoRA fine-tuning is more efficient and less prone to overfitting, so we can train for more steps.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:153]: more efficient should be qualified a bit. My experience is it's slower but more ram and parameter efficient. lower capacity means it's regularized better

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+- We increased the learning rate to 3e-4, but we did not do very thorough hyperparameter tuning. We expect there might be a better learning rate.
+- We found weight decay at 0.1 to lead to better results than no weight decay, so we set it at 0.1.
+- We added the `lora` section to specify the LoRA parameters. All of the parameters are set to the default values.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:156]: maybe move this first?

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+optimizer:
+  learning_rate: 3e-4
+  weight_decay: 0.1
+lora:
+  r: 8  # rank of LoRA transform
+  alpha: 8.0  # scaling factor for LoRA transform
+  dropout: 0.0  # dropout probability for LoRA layers
+```
+
+After training, Levanter will automatically merge the new parameters into the original model and save the new model. We can use it for inference and evaluation.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:179]: this depends on how it's invoked. with `hf_save_path` it will just save the lora adapter that works with PEFT (so just the lora weights plus a pointer), with merged_hf_save_path it will save a merged checkpoint.

Comment (dlwh) [reply docs/examples/Fine-Tuning-Function-Calling.md:179]: it would be good to show the invocation for alpaca_lora because the new lora tutorial doesn't call alpaca_lora

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+## Evaluation
+
+### Metrics
+How do we accurately evaluate a model's performance in function calling tasks?
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:184]: maybe move this earlier?

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+| Metric | Llama2-7B Chat | Full-weight Fine-tuning | LoRA Fine-tuning |
+| ------------------------ | ----- | ----- | ----- |
+| Function Name Accuracy   | 0.014 | 0.577 | 0.517 |
+| Attribute Set Accuracy   | 0.010 | 0.822 | 0.845 |
+| Attribute Value Accuracy | 0.524 | 0.942 | 0.881 |
+| Overall Accuracy         | 0.183 | 0.780 | 0.748 |
+
+There are a few highlights from the results:
+
+- The baseline Llama2-7B Chat model's performance is remarkably low at 0.183 overall accuracy. This is consistent with our earlier observation of its limited capability on this task.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:262]: i'd say it doesn't really know what's expected? "capability" is a bit loaded.

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+- Fine-tuning methods, both full-weight and LoRA, substantially enhance the model's accuracy, achieving 0.780 and 0.748, respectively. Notably, LoRA fine-tuning approaches the performance of full-weight fine-tuning while adjusting only a fraction of the parameters, showcasing its efficiency.
```
Comment (dlwh) [inline docs/examples/Fine-Tuning-Function-Calling.md:263]: might be good to say what the fraction is somewhere, (roughly)

File: mkdocs.yml [hunk 10]
```diff
@@ -88,6 +88,7 @@ nav:
       - 'Configuration-Guide.md'
       - "Fine-Tuning.md"
       - "LoRA.md"
+      - "examples/Fine-Tuning-Function-Calling.md"
```
Comment (dlwh) [inline mkdocs.yml:91]: should we maybe break out an examples section? with Fine-Tuning, LoRA, and this?

**Review by dlwh — COMMENTED** [review]
thanks for putting this together. A few thoughts but otherwise i think this will be helpful

File: docs/examples/Fine-Tuning-Function-Calling.md [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+Expected Response: give_opinion(name[Little Nightmares], rating[good], genres[adventure, platformer, puzzle], player_perspective[side view])
+```
+
+If we test with the Llama2-7B chat model on some examples, we can see its limitation: it struggles to generate accurate function calls and outputs incorrect attributes.
+This highlights the need for fine-tuning: to enhance the model's understanding on this task and producing the intended output.
+
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+Generation from Llama2-7B Chat: \nOutput: inform(name[SpellForce 3], release_year[2017], esrb[NA], genres[fantasy,\naction, adventure], platforms[PC], available_on_steam[no], has_linux_release[no],\nhas_mac_release[no])
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply docs/examples/Fine-Tuning-Function-Calling.md:33]: It actually generate `\n`

File: mkdocs.yml (hunk 10 above)
Comment (Ivan-Zhou) [reply mkdocs.yml:91]: Yes, or `Tutorial` that you added would be great for this.

**Comment by Ivan-Zhou** [comment]
@dlwh Thanks for your comments!

Regarding your comment "is this a good example for function calling" -- yes I agree "semantic parsing" is a more appropriate term for this task, or we should call it "Functional Representation", which I believe is the most precise. 

When I was looking for a dataset, my primary intention was to look for an example of text to json (`Text -> Logical Form`  in your word). Essentially converting text to a structure format that can be parsed in Pydantic/Json. 

Later I found another dataset that is better to fit into the category of "Function Calling": https://huggingface.co/datasets/glaiveai/glaive-function-calling-v2. We can rename the current post as "Functional Representation", and then use the other dataset for another tutorial on Function Calling if needed.

**Comment by dlwh** [comment]
Is functional representation a known technical term for this? I'd prefer to give it a known term of art?

**Comment by Ivan-Zhou** [comment]
@dlwh I have updated the PR based on your comments. A few major changes:
1. I re-position this example from "Function Calling" to "Semantic Parsing". 
2. I re-picked all examples used to describe the task and error analysis; 
3. I added a launching script for LoRA;
4. I added more analysis on the confusion among function names by fine-tuned models;
5. I moved this post to the `Tutorial` section in mkdoc;

File: docs/tutorials/Fine-Tuning-Semantic-Parsing.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+        "composite_score": composite_score,
+    }
+```
+
+### Results
+
+We evaluated the fine-tuned models on a hold-out evaluation set of 714 examples and computed the metrics described above.
+The results are shown in the table below.
+
+| Metric | Llama2-7B Chat | Full-weight Fine-tuning | LoRA Fine-tuning |
```
Comment (dlwh) [inline docs/tutorials/Fine-Tuning-Semantic-Parsing.md:280]: can you transpose this table? it's usually models as rows

File: docs/tutorials/Fine-Tuning-Semantic-Parsing.md [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+
+To expand on the last point, we went deep into predictions of function name.
+The confusion matrices below illustrate how the the full-weight fine-tuning and LoRA fine-tuning models perform in this area.
+
+![Confusion Matrix of Full-weight Fine-Tuning](../figures/finetune_func_cm_full_weight.png)
+![Confusion Matrix of LoRA Fine-Tuning](../figures/finetune_func_cm_lora.png)
+
+- A notable area of confusion is within the `confirm` column, where both models frequently misclassify examples from the `inform`, `recommend`, and `suggest` categories as `confirm`. This may stem from these function names being too similar in the dataset, leading to ambiguity that the models struggle to differentiate.
+- The full-weight fine-tuning model (depicted on the left) tends to make more errors in the "empty" or "other" columns, indicating instances where the model either fails to predict a function name or predicts one that doesn't exist within the predefined set.
+- The LoRA fine-tuning model (shown on the right) exhibits more confusion between the predefined function classes; overall, it has fewer correct predictions along the diagonal of the matrix.
```
Comment (dlwh) [inline docs/tutorials/Fine-Tuning-Semantic-Parsing.md:304]: this is weird to me because it goes against the narrative in my head that LoRA is good for teaching the structure of a task but bad for injecting new information. Do you have any thoughts?

**Review by dlwh — COMMENTED** [review]
sorry i was slow, a few more nits

File: docs/tutorials/Fine-Tuning-Semantic-Parsing.md [hunk 14]
```diff
@@ -0,0 +1,352 @@
+# Fine-Tuning for Semantic Parsing with Levanter
+
+To enable computers to assist human with complex tasks, such as data anlaysis or customer service, they must first understand our natural language and convert it into a format that they can process.
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply docs/tutorials/Fine-Tuning-Semantic-Parsing.md]: I re-wrote the introduction to be more specific on semantic parsing. Hope this version would be more accurate and proper!

File: docs/tutorials/Fine-Tuning-Semantic-Parsing.md (hunk 13 above)
Comment (Ivan-Zhou) [reply docs/tutorials/Fine-Tuning-Semantic-Parsing.md:301]: Yes, the amount of confusion is indeed, bit high. 

I think the results on attributes fields and value does show the advantage of fine-tuning at "teaching about form/structure": both models outputs attributes in the correct format and contains the correct information. 

What I want to show here is that, the selection of the function names is more ambiguous and difficult. It is beyond the "form" problem, but more on reasoning, and the difference between functions `confirm`, `inform`, `suggest` and `recommend` does not seem clear. So I think reducing # functions, clarify the difference, and possibly improve label quality, would be key to improve on the function name accuracy.

File: docs/tutorials/Fine-Tuning-Semantic-Parsing.md [hunk 15]
```diff
@@ -350,3 +347,81 @@ In this post, we showcase the process of fine-tuning a Llama2 model using Levant
 
 Semantic parsing is a critical step in translating user queries into machine-readable, structural format, which is essential for programmatic interactions with AI systems.
 We chose the GEM ViGGO dataset to demonstrate this task. The out-of-box Llama2-7B Chat model struggles to follow the instruction and hallucinates at predicting attributes. By applying fine-tuning, both full-weight and LoRA, we significantly improved the model's ability to perform on this task.
+
+## Appendices
+
+### Appendix A: The Prompt Used in This Task
+
+```
+You are a helpful chatbot to assist users to parse queries in natural language into structural format.
```
Comment (dlwh) [inline docs/tutorials/Fine-Tuning-Semantic-Parsing.md:356]: oh man this is long. I think it's not necessary to use such a long prompt for finetuning (especially I don't think you need few shot examples with fine tuning?)

**Review by dlwh — APPROVED** [review]
lgtm!
